ledger/src/executor.rs

```diff
@@ -0,0 +1,421 @@
+    } else {
+        vec![]
+    };
+
+    let pre_process_units: u64 = aggregate_total_execution_units(timings);
+
+    let (tx_results, balances) = batch.bank().load_execute_and_commit_transactions(
+        batch,
+        MAX_PROCESSING_AGE,
+        transaction_status_sender.is_some(),
+        transaction_status_sender.is_some(),
+        transaction_status_sender.is_some(),
+        transaction_status_sender.is_some(),
+        timings,
+        log_messages_bytes_limit,
+    );
+
+    if bank
+        .feature_set
+        .is_active(&feature_set::gate_large_block::id())
+    {
+        let execution_cost_units = aggregate_total_execution_units(timings) - pre_process_units;
+        let remaining_block_cost_cap = cost_capacity_meter
+            .write()
+            .unwrap()
+            .accumulate(execution_cost_units + tx_cost);
+
+        debug!(
+            "bank {} executed a batch, number of transactions {}, total execute cu {}, total additional cu {}, remaining block cost cap {}",
+            bank.slot(),
+            batch.sanitized_transactions().len(),
+            execution_cost_units,
+            tx_cost,
+            remaining_block_cost_cap,
+        );
+
+        if remaining_block_cost_cap == 0_u64 {
+            return Err(TransactionError::WouldExceedMaxBlockCostLimit);
+        }
+    }
+
+    bank_utils::find_and_send_votes(
+        batch.sanitized_transactions(),
+        &tx_results,
+        replay_vote_sender,
+    );
+
+    let TransactionResults {
+        fee_collection_results,
+        execution_results,
+        rent_debits,
+        ..
+    } = tx_results;
+
+    if let Some(transaction_status_sender) = transaction_status_sender {
+        let transactions = batch.sanitized_transactions().to_vec();
+        let post_token_balances = if record_token_balances {
+            collect_token_balances(bank, batch, &mut mint_decimals)
+        } else {
+            vec![]
+        };
+
+        let token_balances =
+            TransactionTokenBalancesSet::new(pre_token_balances, post_token_balances);
+
+        transaction_status_sender.send_transaction_status_batch(
+            bank.clone(),
+            transactions,
+            execution_results,
+            balances,
+            token_balances,
+            rent_debits,
+            transaction_indexes.to_vec(),
+        );
+    }
+    let first_err = get_first_error(batch, fee_collection_results);
+    first_err.map(|(result, _)| result).unwrap_or(Ok(()))
+}
+
+// The dependency graph contains a set of indices < N that must be executed before index N.
+pub fn build_dependency_graph(
+    tx_account_locks_results: &[Result<TransactionAccountLocks>],
+) -> Result<Vec<HashSet<usize>>> {
+    if let Some(err) = tx_account_locks_results.iter().find(|r| r.is_err()) {
+        err.clone()?;
+    }
+    let transaction_locks: Vec<_> = tx_account_locks_results
+        .iter()
+        .map(|r| r.as_ref().unwrap())
+        .collect();
```

Comment:
You can avoid the double iteration & the `unwrap()` by collecting to a `Result<Vec<T>, E>`
```suggestion
    let transaction_locks = tx_account_locks_results
        .iter()
        .map(|r| r.as_ref())
        .collect::<std::result::Result<Vec<&TransactionAccountLocks>, _>>()
        .map_err(|e| e.clone())?;
```

Reply:
Perhaps taking it a step further and skipping these 2 full iterations entirely by changing the format of the next one as:
```rust
    for (idx, tx_account_locks) in transaction_locks.iter().enumerate() {
        let tx_account_locks = tx_account_locks.as_ref()?;
        
        for account in &tx_account_locks.readonly {
            // ...
        }
        for account in &tx_account_locks.writable {
            // ...
        }
    }
```